diff --git a/codegen.c b/codegen.c
index fd6993b..6ec8da7 100644
--- a/codegen.c
+++ b/codegen.c
@@ -465,11 +465,12 @@ static void emit_data(Obj *prog) {
     if (var->is_function)
       continue;
 
-    println("  .data");
     println("  .globl %s", var->name);
-    println("%s:", var->name);
 
     if (var->init_data) {
+      println("  .data");
+      println("%s:", var->name);
+
       Relocation *rel = var->rel;
       int pos = 0;
       while (pos < var->ty->size) {
@@ -481,9 +482,12 @@ static void emit_data(Obj *prog) {
           println("  .byte %d", var->init_data[pos++]);
         }
       }
-    } else {
-      println("  .zero %d", var->ty->size);
+      continue;
     }
+
+    println("  .bss");
+    println("%s:", var->name);
+    println("  .zero %d", var->ty->size);
   }
 }
 
