diff --git a/parse.c b/parse.c
index eb7d1bc..521013c 100644
--- a/parse.c
+++ b/parse.c
@@ -455,9 +455,14 @@ static Type *declspec(Token **rest, Token *tok, VarAttr *attr) {
   return ty;
 }
 
-// func-params = (param ("," param)*)? ")"
+// func-params = ("void" | param ("," param)*)? ")"
 // param       = declspec declarator
 static Type *func_params(Token **rest, Token *tok, Type *ty) {
+  if (equal(tok, "void") && equal(tok->next, ")")) {
+    *rest = tok->next->next;
+    return func_type(ty);
+  }
+
   Type head = {};
   Type *cur = &head;
 
diff --git a/test/function.c b/test/function.c
index d3821cc..72a3715 100644
--- a/test/function.c
+++ b/test/function.c
@@ -1,6 +1,6 @@
 #include "test.h"
 
-int ret3() {
+int ret3(void) {
   return 3;
   return 5;
 }
@@ -41,7 +41,7 @@ int sub_short(short a, short b, short c) {
 
 int g1;
 
-int *g1_ptr() { return &g1; }
+int *g1_ptr(void) { return &g1; }
 char int_to_char(int x) { return x; }
 
 int div_long(long a, long b) {
@@ -51,7 +51,7 @@ int div_long(long a, long b) {
 _Bool bool_fn_add(_Bool x) { return x + 1; }
 _Bool bool_fn_sub(_Bool x) { return x - 1; }
 
-static int static_fn() { return 3; }
+static int static_fn(void) { return 3; }
 
 int param_decay(int x[]) { return x[0]; }
 
