diff --git a/parse.c b/parse.c
index 26711fc..6f689fd 100644
--- a/parse.c
+++ b/parse.c
@@ -516,6 +516,9 @@ static Type *func_params(Token **rest, Token *tok, Type *ty) {
     cur = cur->next = copy_type(ty2);
   }
 
+  if (cur == &head)
+    is_variadic = true;
+
   ty = func_type(ty);
   ty->params = head.next;
   ty->is_variadic = is_variadic;
@@ -2049,6 +2052,9 @@ static Node *funcall(Token **rest, Token *tok) {
     Node *arg = assign(&tok, tok);
     add_type(arg);
 
+    if (!param_ty && !ty->is_variadic)
+      error_tok(tok, "too many arguments");
+
     if (param_ty) {
       if (param_ty->kind == TY_STRUCT || param_ty->kind == TY_UNION)
         error_tok(arg->tok, "passing struct or union is not supported yet");
@@ -2059,6 +2065,9 @@ static Node *funcall(Token **rest, Token *tok) {
     cur = cur->next = arg;
   }
 
+  if (param_ty)
+    error_tok(tok, "too few arguments");
+
   *rest = skip(tok, ")");
 
   Node *node = new_node(ND_FUNCALL, start);
