diff --git a/main.c b/main.c
index 4abfeca..8fcc1a3 100644
--- a/main.c
+++ b/main.c
@@ -29,6 +29,17 @@ static bool take_arg(char *arg) {
   return false;
 }
 
+static void add_default_include_paths(char *argv0) {
+  // We expect that chibicc-specific include files are installed
+  // to ./include relative to argv[0].
+  strarray_push(&include_paths, format("%s/include", dirname(strdup(argv0))));
+
+  // Add standard include paths.
+  strarray_push(&include_paths, "/usr/local/include");
+  strarray_push(&include_paths, "/usr/include/x86_64-linux-gnu");
+  strarray_push(&include_paths, "/usr/include");
+}
+
 static void parse_args(int argc, char **argv) {
   // Make sure that all command line options that take an argument
   // have an argument.
@@ -313,6 +324,7 @@ int main(int argc, char **argv) {
   parse_args(argc, argv);
 
   if (opt_cc1) {
+    add_default_include_paths(argv[0]);
     cc1();
     return 0;
   }
diff --git a/self.py b/self.py
index 88af153..6dd9746 100755
--- a/self.py
+++ b/self.py
@@ -71,6 +71,7 @@ int strncmp(char *p, char *q, long n);
 void *memcpy(char *dst, char *src, long n);
 char *strdup(char *p);
 char *strndup(char *p, long n);
+char *strdup(char *p);
 int isspace(int c);
 int ispunct(int c);
 int isdigit(int c);
