diff --git a/parse.c b/parse.c
index 98389e3..bcfab61 100644
--- a/parse.c
+++ b/parse.c
@@ -2398,6 +2398,13 @@ static Token *function(Token *tok, Type *basety, VarAttr *attr) {
     fn->va_area = new_lvar("__va_area__", array_of(ty_char, 136));
 
   tok = skip(tok, "{");
+
+  // [https://www.sigbus.info/n1570#6.4.2.2p1] "__func__" is
+  // automatically defined as a local variable containing the
+  // current function name.
+  push_scope("__func__")->var =
+    new_string_literal(fn->name, array_of(ty_char, strlen(fn->name) + 1));
+
   fn->body = compound_stmt(&tok, tok);
   fn->locals = locals;
   leave_scope();
diff --git a/test/function.c b/test/function.c
index 64de04c..cdc950c 100644
--- a/test/function.c
+++ b/test/function.c
@@ -114,6 +114,10 @@ int (*fnptr(int (*fn)(int n, ...)))(int, ...) {
 
 int param_decay2(int x()) { return x(); }
 
+char *func_fn(void) {
+  return __func__;
+}
+
 int main() {
   ASSERT(3, ret3());
   ASSERT(8, add2(3, 5));
@@ -191,6 +195,10 @@ int main() {
 
   ASSERT(3, param_decay2(ret3));
 
+  ASSERT(5, sizeof(__func__));
+  ASSERT(0, strcmp("main", __func__));
+  ASSERT(0, strcmp("func_fn", func_fn()));
+
   printf("OK\n");
   return 0;
 }
