diff --git a/parse.c b/parse.c
index 633bf97..384a1b0 100644
--- a/parse.c
+++ b/parse.c
@@ -2003,8 +2003,13 @@ static Node *unary(Token **rest, Token *tok) {
   if (equal(tok, "-"))
     return new_unary(ND_NEG, cast(rest, tok->next), tok);
 
-  if (equal(tok, "&"))
-    return new_unary(ND_ADDR, cast(rest, tok->next), tok);
+  if (equal(tok, "&")) {
+    Node *lhs = cast(rest, tok->next);
+    add_type(lhs);
+    if (lhs->kind == ND_MEMBER && lhs->member->is_bitfield)
+      error_tok(tok, "cannot take address of bitfield");
+    return new_unary(ND_ADDR, lhs, tok);
+  }
 
   if (equal(tok, "*")) {
     // [https://www.sigbus.info/n1570#6.5.3.2p4] This is an oddity
