diff --git a/tokenize.c b/tokenize.c
index 1453929..3fd5a1b 100644
--- a/tokenize.c
+++ b/tokenize.c
@@ -562,6 +562,25 @@ File *new_file(char *name, int file_no, char *contents) {
   return file;
 }
 
+// Replaces \r or \r\n with \n.
+static void canonicalize_newline(char *p) {
+  int i = 0, j = 0;
+
+  while (p[i]) {
+    if (p[i] == '\r' && p[i + 1] == '\n') {
+      i += 2;
+      p[j++] = '\n';
+    } else if (p[i] == '\r') {
+      i++;
+      p[j++] = '\n';
+    } else {
+      p[j++] = p[i++];
+    }
+  }
+
+  p[j] = '\0';
+}
+
 // Removes backslashes followed by a newline.
 static void remove_backslash_newline(char *p) {
   int i = 0, j = 0;
@@ -594,6 +613,7 @@ Token *tokenize_file(char *path) {
   if (!p)
     return NULL;
 
+  canonicalize_newline(p);
   remove_backslash_newline(p);
 
   // Save the filename for assembler .file directive.
