diff --git a/parse.c b/parse.c
index 77144d7..ce97ba1 100644
--- a/parse.c
+++ b/parse.c
@@ -875,7 +875,7 @@ static int array_designator(Token **rest, Token *tok, Type *ty) {
   return i;
 }
 
-// designation = ("[" const-expr "]")* "=" initializer
+// designation = ("[" const-expr "]")* "="? initializer
 static void designation(Token **rest, Token *tok, Initializer *init) {
   if (equal(tok, "[")) {
     if (init->ty->kind != TY_ARRAY)
@@ -886,7 +886,8 @@ static void designation(Token **rest, Token *tok, Initializer *init) {
     return;
   }
 
-  tok = skip(tok, "=");
+  if (equal(tok, "="))
+    tok = tok->next;
   initializer2(rest, tok, init);
 }
 
diff --git a/test/initializer.c b/test/initializer.c
index c13e379..c7d3ed4 100644
--- a/test/initializer.c
+++ b/test/initializer.c
@@ -213,6 +213,9 @@ int main() {
   ASSERT(0, strcmp(g65.b, "oo"));
   ASSERT(0, strcmp(g66.b, "oobar"));
 
+  ASSERT(7, ((int[10]){ [3] 7 })[3]);
+  ASSERT(0, ((int[10]){ [3] 7 })[4]);
+
   printf("OK\n");
   return 0;
 }
