diff --git a/preprocess.c b/preprocess.c
index 7b40774..dbe74e4 100644
--- a/preprocess.c
+++ b/preprocess.c
@@ -907,6 +907,10 @@ static Token *timestamp_macro(Token *tmpl) {
   return new_str_token(buf, tmpl);
 }
 
+static Token *base_file_macro(Token *tmpl) {
+  return new_str_token(base_file, tmpl);
+}
+
 // __DATE__ is expanded to the current date, e.g. "May 17 2020".
 static char *format_date(struct tm *tm) {
   static char mon[][4] = {
@@ -972,6 +976,7 @@ void init_macros(void) {
   add_builtin("__LINE__", line_macro);
   add_builtin("__COUNTER__", counter_macro);
   add_builtin("__TIMESTAMP__", timestamp_macro);
+  add_builtin("__BASE_FILE__", base_file_macro);
 
   time_t now = time(NULL);
   struct tm *tm = localtime(&now);
diff --git a/test/macro.c b/test/macro.c
index 1f8f13d..be6fe71 100644
--- a/test/macro.c
+++ b/test/macro.c
@@ -375,6 +375,8 @@ int main() {
 
   ASSERT(24, strlen(__TIMESTAMP__));
 
+  ASSERT(0, strcmp(__BASE_FILE__, "test/macro.c"));
+
   printf("OK\n");
   return 0;
 }
