diff --git a/preprocess.c b/preprocess.c
index e0d3281..962ad3b 100644
--- a/preprocess.c
+++ b/preprocess.c
@@ -996,7 +996,6 @@ void init_macros(void) {
   define_macro("__STDC_HOSTED__", "1");
   define_macro("__STDC_NO_ATOMICS__", "1");
   define_macro("__STDC_NO_COMPLEX__", "1");
-  define_macro("__STDC_NO_VLA__", "1");
   define_macro("__STDC_UTF_16__", "1");
   define_macro("__STDC_UTF_32__", "1");
   define_macro("__STDC_VERSION__", "201112L");
