diff --git a/codegen.c b/codegen.c
index 3c002f0..42e9e29 100644
--- a/codegen.c
+++ b/codegen.c
@@ -250,6 +250,7 @@ static void assign_lvar_offsets(Obj *prog) {
     int offset = 0;
     for (Obj *var = fn->locals; var; var = var->next) {
       offset += var->ty->size;
+      offset = align_to(offset, var->ty->align);
       var->offset = -offset;
     }
     fn->stack_size = align_to(offset, 16);
diff --git a/test/variable.c b/test/variable.c
index 2c995fd..0e95391 100644
--- a/test/variable.c
+++ b/test/variable.c
@@ -47,6 +47,9 @@ int main() {
   ASSERT(2, ({ int x=2; { int x=3; } int y=4; x; }));
   ASSERT(3, ({ int x=2; { x=3; } x; }));
 
+  ASSERT(15, ({ int x; int y; char z; char *a=&y; char *b=&z; b-a; }));
+  ASSERT(1, ({ int x; char y; int z; char *a=&y; char *b=&z; b-a; }));
+
   printf("OK\n");
   return 0;
 }
